src/System.Private.CoreLib/shared/System/Text/AsciiUtility.cs

```diff
@@ -0,0 +1,1655 @@
+            }
+
+            // Try reading 16 bits.
+
+            if (((uint)bufferLength & 2) != 0)
+            {
+                currentUInt32 = Unsafe.ReadUnaligned<ushort>(ref buffer);
+                if (!AllBytesInUInt32AreAscii(currentUInt32))
+                {
+                    goto FoundNonAsciiData;
+                }
+
+                buffer = ref Unsafe.Add(ref buffer, 2);
+            }
+
+            // Try reading 8 bits
+
+            if (((uint)bufferLength & 1) != 0)
+            {
+                // If the buffer contains non-ASCII data, the comparison below will fail, and
+                // we'll end up not incrementing the buffer reference.
+
+                if ((sbyte)buffer >= 0)
+                {
+                    buffer = ref Unsafe.Add(ref buffer, 1);
+                }
+            }
+
+        Finish:
+
+            // It's possible that the new buffer reference could point _just beyond_ the original buffer,
+            // but this is safe from a GC-tracking perspective.
+
+            return (nuint)(void*)Unsafe.ByteOffset(ref originalBuffer, ref buffer);
+
+        FoundNonAsciiData:
+
+            Debug.Assert(!AllBytesInUInt32AreAscii(currentUInt32), "Shouldn't have reached this point if we have an all-ASCII input.");
+
+            // The method being called doesn't bother looking at whether the high byte is ASCII. There are only
+            // two scenarios: (a) either one of the earlier bytes is not ASCII and the search terminates before
+            // we get to the high byte; or (b) all of the earlier bytes are ASCII, so the high byte must be
+            // non-ASCII. In both cases we only care about the low 24 bits.
+
+            buffer = ref Unsafe.AddByteOffset(ref buffer, CountNumberOfLeadingAsciiBytesFrom24BitInteger(currentUInt32));
+            goto Finish;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveOptimization)]
+        private static unsafe nuint GetIndexOfFirstNonAsciiByte_Sse2(ref byte buffer, nuint bufferLength)
+        {
+            // Quick check for empty inputs.
+
+            if (bufferLength == 0)
+            {
+                return 0;
+            }
+
+            // JIT turns the below into constants
+
+            uint SizeOfVector128 = (uint)Unsafe.SizeOf<Vector128<byte>>();
+            nuint MaskOfAllBitsInVector128 = (nuint)(SizeOfVector128 - 1);
+
+            Debug.Assert(Sse2.IsSupported, "Should've been checked by caller.");
+            Debug.Assert(BitConverter.IsLittleEndian, "SSE2 assumes little-endian.");
+
+            // For best performance this method performs all-aligned vector reads. This means that
+            // we could end up reading data before or after the desired buffer, but we'll mask it
+            // out when performing the "is ASCII?" check. This is guaranteed not to AV since such reads
+            // will always be in the same page as the actual buffer itself.
```

Comment:
I don't think the performance loss would be that great, especially given it won't be crossing a page boundary, if this did at most 2 unaligned reads (one at the start to become aligned and another at the end to handle trailing elements). There would still be a hit in the case either of those reads crosses a cache-line boundary, but that should probably be fine.

Reply:
That might also remove anyone's concerns about this purposefully overrunning the buffer.

Reply:
I tried this in an earlier iteration, but it caused significant perf regressions for small inputs. Perhaps with some other changes that were made since then it's not quite as impactful as it was early on. I'll rerun the numbers.

Reply:
Hmm, I actually wouldn't have expected that given that you already have the branches for dealing with leading/trailing masks and you are basically just changing the logic to start from index 0 and add x to become aligned; rather than starting from index 0 and subtracting y to become aligned.

Reply:
IIRC the slowdown occurred because the code had to handle the case where there was less than a single vectors worth of data total. So try read a qword, then a dword, then a word, then a byte. Lots of memory accesses compared to a single aligned vector read (which per Agner has the same latency and throughput as a normal q/dword mov from mem to register).

Reply:
Ah, that would make sense.